### PR TITLE
[SYCL][FPGA] Update [[intel::max_concurrency()]] attribute support

### PR DESCRIPTION
**Pull request by smanna12**
This patch improves support for [[intel::max_concurrency()]]  implementation.
   1. Fixes bugs the way attribute handled duplicate attribute diagnostics.
   2. Adds missing diagnostics test cases.
   3. Splits AST and diagnostics part for better codes readability(part of https://github.com/intel/llvm/issues/4118).
   
Signed-off-by: Soumi Manna <soumi.manna@intel.com>

### CODE REVIEW

**Review by premanandrao — DISMISSED** [review]
LGTM

**Comment by smanna12** [comment]
Than you for reviews.

**Comment by smanna12** [comment]
@intel/llvm-gatekeepers, this PR is ready to be merged.

The failure does not related to my PR:
http://icl-jenkins.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FLin%2FBasic_E2E/detail/Basic_E2E/2517/pipeline

Failed Tests (1):
SYCL :: Basic/queue/release.cpp

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 1]
```diff
@@ -7623,14 +7623,17 @@ SYCLIntelFPGAMaxConcurrencyAttr *Sema::MergeSYCLIntelFPGAMaxConcurrencyAttr(
...
-      Diag(A.getLoc(), diag::note_previous_attribute);
+    if (const auto *DeclExpr =
+            dyn_cast<ConstantExpr>(DeclAttr->getNThreadsExpr())) {
+      if (const auto *MergeExpr = dyn_cast<ConstantExpr>(A.getNThreadsExpr())) {
+        if (DeclExpr->getResultAsAPSInt() != MergeExpr->getResultAsAPSInt()) {
+          Diag(DeclAttr->getLoc(), diag::warn_duplicate_attribute) << &A;
+          Diag(A.getLoc(), diag::note_previous_attribute);
+        }
+        // Do not add a duplicate attribute.
+        return nullptr;
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaDeclAttr.cpp:7634]: I'm not sure why this change was made. I feel the old code was  easier to read. You've also slightly changed behavior here right? The return was hit before if there was an existing attribute. Now the return is inside - 

```
 if (const auto *DeclExpr =
            dyn_cast<ConstantExpr>(DeclAttr->getNThreadsExpr()))
```

Comment (elizabethandrews) [inline clang/lib/Sema/SemaDeclAttr.cpp:7634]: I'm not sure why this change was made. I feel the old code was  easier to read. You've also slightly changed behavior here right? The return was hit before if there was an existing attribute. Now the return is inside - 

```
 if (const auto *DeclExpr =
            dyn_cast<ConstantExpr>(DeclAttr->getNThreadsExpr()))
```

**Comment by pvchupin** [comment]
@smanna12 can you please sync with @elizabethandrews on recent concerns?

File: clang/test/SemaSYCL/max-concurrency-ast.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+template <int N>
+class KernelFunctor2 {
+public:
+  [[intel::max_concurrency(N)]] void operator()() const {
+  }
+};
+
+void foo() {
+  q.submit([&](handler &h) {
+    // CHECK: FunctionDecl {{.*}}kernel_name_1
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/max-concurrency-ast.cpp:59]: Can you explain what you are testing here?  Is this attribute supposed to be propagated to kernel or not ?

File: clang/test/SemaSYCL/max-concurrency-ast.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        []() [[intel::max_concurrency(4)]] {});
+
+    // Ignore duplicate attribute.
+    h.single_task<class kernel_name_4>(
+        // CHECK: FunctionDecl {{.*}}kernel_name_4
+        // CHECK: SYCLIntelFPGAMaxConcurrencyAttr
+        // CHECK-NEXT: ConstantExpr {{.*}} 'int'
+        // CHECK-NEXT: value: Int 3
+        // CHECK-NEXT: IntegerLiteral{{.*}}3{{$}}
+        []() [[intel::max_concurrency(3),
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/max-concurrency-ast.cpp:89]: Please add a CHECK-NOT here to verify there is no duplicate added.

File: clang/test/SemaSYCL/max-concurrency.cpp [hunk 4]
```diff
@@ -1,159 +1,86 @@
...
+// No diagnostic is emitted because the arguments match.
+[[intel::max_concurrency(4)]] void func4();
+[[intel::max_concurrency(4)]] void func4(); // OK
+
+// Diagnostic is emitted because the arguments mismatch.
+[[intel::max_concurrency(2)]] // expected-note {{previous attribute is here}}
+[[intel::max_concurrency(4)]] void
+func5() {} // expected-warning@-1 {{attribute 'max_concurrency' is already applied with different arguments}}
+
+[[intel::max_concurrency(1)]] void func6(); // expected-note {{previous attribute is here}}
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/max-concurrency.cpp:29]: This is same test as `func5 `right?

File: clang/test/SemaSYCL/max-concurrency-ast.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (smanna12) [reply clang/test/SemaSYCL/max-concurrency-ast.cpp:60]: Yes, test is checking that the attribute does not get propagated here.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 5]
```diff
@@ -7654,14 +7657,21 @@ void Sema::AddSYCLIntelFPGAMaxConcurrencyAttr(Decl *D,
...
+      // have converted it to a constant expression yet and thus we test
+      // whether this is a null pointer.
+      if (const auto *DeclExpr =
+              dyn_cast<ConstantExpr>(DeclAttr->getNThreadsExpr())) {
+        if (ArgVal != DeclExpr->getResultAsAPSInt()) {
+          Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+          Diag(DeclAttr->getLoc(), diag::note_previous_attribute);
+        }
+        // Drop the duplicate attribute.
+        return;
```
[the remark this answers is not in the document]
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:7673]: The 'return' statement has been moved so that we do not add duplicate attribute here:
Otherwise we will have duplicate attribute on AST:

```
|-OpenCLKernelAttr 0x55af79b232a0 <<invalid sloc>> Implicit
  |-ArtificialAttr 0x55af79b232f8 <<invalid sloc>> Implicit
  |-SYCLKernelAttr 0x55af79b234e8 <<invalid sloc>> Implicit
  |-AsmLabelAttr 0x55af79b23cb0 <<invalid sloc>> Implicit "_ZTSZZ3foovENKUlRN2cl4sycl7handlerEE_clES2_E13kernel_name_4"
  |-SYCLIntelFPGAMaxConcurrencyAttr 0x55af79b1ad60 <llvm/clang/test/SemaSYCL/max-concurrency-ast.cpp:90:16, col:40>
  | `-ConstantExpr 0x55af79b1ad40 <col:39> 'int'
  |   |-value: Int 3
  |   `-IntegerLiteral 0x55af79b1aab8 <col:39> 'int' 3
  `-SYCLIntelFPGAMaxConcurrencyAttr 0x55af79b1ade0 <line:91:16, col:40>
    `-ConstantExpr 0x55af79b1adc0 <col:39> 'int'
      |-value: Int 3
      `-IntegerLiteral 0x55af79b1aad8 <col:39> 'int' 3
```

[the remark this answers is not in the document]
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:7673]: Same comment above

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 1 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:7634]: Sorry for late response. Thanks @elizabethandrews for reviews.

I found a bug with the original implementation while working on splitting tests (AST vs Sema tests).

Template case(below) for duplicate attributes does not work well with the original implementation: Test compiles fine without any diagnostic. I have made similar changes like other function attributes to handle template cases for duplicate attribute values with max_concurrency attribute properly. 

```
template <int size>
[[intel::max_concurrency(10)]] void func11(); // expected-note {{previous attribute is here}}
template <int size>
[[intel::max_concurrency(size)]] void func11() {} // expected-warning {{attribute 'max_concurrency' is already applied with different arguments}}

void checkTemplates() {
  
  func11<20>(); // expected-note {{in instantiation of function template specialization 'func11<20>' requested here}}
}
```

**Review by elizabethandrews — APPROVED** [review]
LGTM Thanks!

**Comment by smanna12** [comment]
The failures are not related to my change:

Failed Tests (6):
  SYCL :: GroupAlgorithm/SYCL2020/exclusive_scan.cpp
  SYCL :: GroupAlgorithm/SYCL2020/inclusive_scan.cpp
  SYCL :: GroupAlgorithm/SYCL2020/reduce.cpp
  SYCL :: GroupAlgorithm/exclusive_scan.cpp
  SYCL :: GroupAlgorithm/inclusive_scan.cpp
  SYCL :: GroupAlgorithm/reduce.cpp

**Comment by smanna12** [comment]
Thank you for the reviews.

**Comment by smanna12** [comment]
@intel/llvm-gatekeepers, this PR is ready to be merged. Thank you

**Comment by pvchupin** [comment]
Fails are handled at https://github.com/intel/llvm-test-suite/pull/1034
